### PR TITLE
Export stock list

### LINKED ISSUE #1587: View Stock: add a button "Export" to down the stock data on an excel or csv format
## Describe the solution you'd like 🎁

Add an export button in the "View Stock" menu to download stock data on an excel format

Screen Shot 2023-04-20 at 4 47 34 PM

Would also be nice to have a print button to download a report in a pdf format

### Moneyworks Jobcode 🧰

DJUNCUST

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1587 

# 👩🏻‍💻 What does this PR do? 
 
Adds the export button to the stock list view

# 🧪 How has/should this change been tested? 

Go to the stock list view: there's now an export button.
Clicking it will export 😉 

## 💌 Any notes for the reviewer?

I now wonder if the button should be disabled if there's no items in the list.. but that would be inconsistent, so let's not do that right now

## 📃 Documentation

- [ ] Update screenshots?
- [ ] Export functionality

### CODE REVIEW

File: client/packages/system/src/Stock/ListView/AppBarButtons.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+} from '@openmsupply-client/common';
+import { StockLineRowFragment, useStock } from '../api';
+import { stockLinesToCsv } from '../../utils';
+
+export const AppBarButtonsComponent: FC<{
+  sortBy: SortBy<StockLineRowFragment>;
+}> = ({ sortBy }) => {
+  const { success, error } = useNotification();
+  const t = useTranslation(['distribution', 'common']);
+  const { fetchAsync, isLoading } = useStock.line.listAll(sortBy);
```
Comment (CarlosNZ) [inline client/packages/system/src/Stock/ListView/AppBarButtons.tsx:22]: Small thing, not sure what the correct answer is, but I'm not sure the "Export" should necessarily use the sort order of the current view. It doesn't respect other aspects of the current display (such as the filtering), so it's possibly inconsistent/misleading to export something different depending on the current displayed sort.

I would think that just exporting by "name" sorting would be okay. It also means that if you make different exports on different days, you can compare your lists much easier. And if they need any additional sorting/filtering they can just their spreadsheet app for that.

File: client/packages/system/src/Stock/ListView/AppBarButtons.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  return (
+    <AppBarButtonsPortal>
+      <Grid container gap={1}>
+        <LoadingButton
+          startIcon={<DownloadIcon />}
+          isLoading={isLoading}
+          variant="outlined"
+          onClick={csvExport}
+          disabled={EnvUtils.platform === Platform.Android}
```
Comment (CarlosNZ) [inline client/packages/system/src/Stock/ListView/AppBarButtons.tsx:44]: So it's disabled on Android? Might it be better to just hide it completely in that case?

File: client/packages/system/src/utils.ts [hunk 3]
```diff
@@ -42,3 +43,37 @@ export const masterListsToCsv = (
...
+    t('label.available-packs'),
+    t('label.supplier'),
+  ];
+
+  const data = invoices.map(node => [
+    node.id,
+    node.item.code,
+    node.item.name,
+    node.batch,
+    Formatter.csvDateString(node.expiryDate),
```
Comment (CarlosNZ) [inline client/packages/system/src/utils.ts:70]: This includes a time in the output date string, e.g. `2024-07-27 12:00:00`. Is this what we want? (Wouldn't just the date be more appropriate?)

**Review by CarlosNZ — COMMENTED** [review]
Looks good, working nicely. Just a couple of small questions/suggestions.

File: client/packages/system/src/Stock/ListView/AppBarButtons.tsx (hunk 1 above)
Comment (mark-prins) [reply client/packages/system/src/Stock/ListView/AppBarButtons.tsx:22]: it seemed like a cute idea at the time - though like you say, the pagination and filtering isn't used when exporting so it does make the sort inconsistent. I'll run through and update them all.

File: client/packages/system/src/Stock/ListView/AppBarButtons.tsx (hunk 2 above)
Comment (mark-prins) [reply client/packages/system/src/Stock/ListView/AppBarButtons.tsx:44]: For UI items, the preference is for functionality which isn't available to still be shown - it tells the user that they can do something.. as long as they have the right platform / permissions etc. I prefer having an uncluttered UI, but can see the point.

The android fix was a quick and dirty one - with a plan to revisit and implement later. I'd prefer to keep the button visible, as a reminder to make it work sometime.

File: client/packages/system/src/utils.ts (hunk 3 above)
Comment (mark-prins) [reply client/packages/system/src/utils.ts:70]: I'll add an expiryDateString helper..

**Comment by mark-prins** [comment]
Thanks @CarlosNZ - the comment about sorting has resulted in a few bug fixes along the way, and the sort is much neater now!

**Review by CarlosNZ — APPROVED** [review]
Great, thanks for the changes 👍

**Pull request merged**